src/main/scala/xiangshan/frontend/ifu/Ifu.scala

```diff
@@ -454,37 +477,60 @@ class Ifu(implicit p: Parameters) extends IfuModule
       }
   }
 
-  val s2_firstTakenIdx = WireDefault(0.U.asTypeOf(ValidUndirectioned(UInt(log2Ceil(PredictWidth).W))))
-  s2_firstTakenIdx.valid := firstFtqOffset.valid
-  s2_firstTakenIdx.bits := Mux(
-    firstFtqOffset.valid,
-    instrCountBeforeCurrent(firstFtqOffset.bits + 1.U(5.W)) - 1.U,
-    PopCount(bubbleInstrValid.asUInt & s2_firstInstrRange) - 1.U
-  )
-
-  private val s2_firstBlock  = WireDefault(0.U.asTypeOf(new FetchBlockInfo))
-  private val s2_secondBlock = WireDefault(0.U.asTypeOf(new FetchBlockInfo))
-  s2_firstBlock.ftqIdx       := s2_ftqReq.ftqIdx
-  s2_firstBlock.predTakenIdx := s2_firstTakenIdx
-  s2_firstBlock.invalidTaken := !bubbleInstrValid(firstFtqOffset.bits) && firstFtqOffset.valid
-
-  s2_firstBlock.instrRange := s2_firstInstrRange // Temporary compromise for V2 compatibility; actually uses s1_firstJumpRange & s1_firstFtrRange
-  s2_firstBlock.pcHigh           := s2_ftqReq.startAddr(VAddrBits - 1, PcCutPoint)
-  s2_firstBlock.pcHighPlus1      := s2_ftqReq.startAddr(VAddrBits - 1, PcCutPoint) + 1.U
-  s2_firstBlock.target           := s2_ftqReq.nextStartAddr
-  s2_firstBlock.fetchSize        := s2_firstFetchSize
-  s2_firstBlock.bubbleInstrValid := bubbleInstrValid.asUInt & s2_firstInstrRange
+  val s2_fetchTakenIdx = Wire(Vec(FetchPorts, ValidUndirectioned(UInt(log2Ceil(PredictWidth).W))))
+  for (i <- 0 until FetchPorts) {
+    s2_fetchTakenIdx(i).valid := s2_ftqFetch(i).ftqOffset.valid
+    s2_fetchTakenIdx(i).bits  := PopCount(rawInstrValid.asUInt & s2_instrRange(i)) - 1.U
+  }
 
-  s2_secondBlock.bubbleInstrValid := (bubbleInstrValid.asUInt >> s2_firstFetchSize) & s2_secondInstrRange
-  private val s2_rawDataDupWire = VecInit(Seq.fill(2)(s2_rawData))
+  s2_fetchTakenIdx(0).valid := s2_ftqFetch(0).ftqOffset.valid && s2_firstValid
+  s2_fetchTakenIdx(1).valid := s2_ftqFetch(1).ftqOffset.valid && s2_secondValid
+
+  private val s2_fetchBlock = WireDefault(0.U.asTypeOf(Vec(FetchPorts, new FetchBlockInfo)))
+  for (i <- 0 until FetchPorts) {
+    s2_fetchBlock(i).ftqIdx          := s2_ftqFetch(i).ftqIdx
+    s2_fetchBlock(i).doubline        := s2_doubleline(i)
+    s2_fetchBlock(i).predTakenIdx    := s2_fetchTakenIdx(i)
+    s2_fetchBlock(i).invalidTaken    := !rawInstrValid(s2_ftqFetch(i).ftqOffset.bits) && s2_ftqFetch(i).ftqOffset.valid
+    s2_fetchBlock(i).ftqOffset.valid := s2_ftqFetch(i).ftqOffset.valid
+    s2_fetchBlock(i).ftqOffset.bits  := s2_ftqFetch(i).ftqOffset.bits
+    s2_fetchBlock(i).instrRange      := s2_instrRange(i)
+    s2_fetchBlock(i).pcHigh          := s2_ftqFetch(i).startAddr(VAddrBits - 1, PcCutPoint)
+    s2_fetchBlock(i).pcHighPlus1     := s2_ftqFetch(i).startAddr(VAddrBits - 1, PcCutPoint) + 1.U
+    s2_fetchBlock(i).startAddr       := s2_ftqFetch(i).startAddr
+    s2_fetchBlock(i).target          := s2_ftqFetch(i).nextStartAddr
+    s2_fetchBlock(i).fetchSize       := s2_fetchSize(i)
+    s2_fetchBlock(i).rawInstrValid   := rawInstrValid.asUInt & s2_instrRange(i)
+  }
```

Comment:
```suggestion
  private val s2_fetchBlock = VecInit((0 until FetchPorts).map{ i =>
    val b = Wire(new FetchBlock)
    b.ftqIdx          := s2_ftqFetch(i).ftqIdx
    b.doubline        := s2_doubleline(i)
    b.predTakenIdx    := s2_fetchTakenIdx(i)
    b.invalidTaken    := !rawInstrValid(s2_ftqFetch(i).ftqOffset.bits) && s2_ftqFetch(i).ftqOffset.valid
    b.ftqOffset.valid := s2_ftqFetch(i).ftqOffset.valid
    b.ftqOffset.bits  := s2_ftqFetch(i).ftqOffset.bits
    b.instrRange      := s2_instrRange(i)
    b.pcHigh          := s2_ftqFetch(i).startAddr(VAddrBits - 1, PcCutPoint)
    b.pcHighPlus1     := s2_ftqFetch(i).startAddr(VAddrBits - 1, PcCutPoint) + 1.U
    b.startAddr       := s2_ftqFetch(i).startAddr
    b.target          := s2_ftqFetch(i).nextStartAddr
    b.fetchSize       := s2_fetchSize(i)
    b.rawInstrValid   := rawInstrValid.asUInt & s2_instrRange(i)
  })
```

It's more recommended to simply using a `VecInit((...).map{ <assignments> })` instead of `WireDefault` + `for (...) { <assignments> }`